power: Documentation updates and general fixes

ceolin opened the pull request:
- Restore IRQ when post ops is not implemented by the target
 - Move PM hooks to their documentation section
 - Add notes about notifier constraints
 - Removed reference for undefined variable
 - Send exit notification after exit_post_ops callback

Commit:
power: Update notifier documentation

Add a new note clarifying one constraint in this API.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Remove unused symbol

pm_idle_exit_notify is no longer defined or used. Just removing it.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Fix hooks documentation

Put documentation about pm_power_state_set in the proper group,
power_management_hook_interface.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: pm_power_state_exit_post_ops has to unlock irqs

The kernel expects irqs be unlocked after this function be called.
Add it to the weak function in case of the SoC or application has not
implemented it.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Avoid duplicated code

Just call the function instead of replicate the logic.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
power: Send exit notification after post_ops callback

Currently notification about exiting an idle state is sent before
post_ops and with interruptions locked. Change it to be sent after
exit_pos_ops callback be called.

Unfortunately it is not possible to just change the order these
functions are called since the idle thread can be scheduled out just
after irqs be unlocked and before has the chance to send the
notification, so we are locking the scheduler and unlocking it only
after the notification be sent.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
doc: power: General improvements

Add information about power state constraints and more information
about power management policy.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

erwango commented on include/power/power.h:
```diff
@@ -48,6 +48,10 @@ extern unsigned char pm_idle_exit_notify;
  * @note These callbacks can be called from the ISR of the event
  *       that caused the kernel exit from idling.
+ *
+ * @note It is not allowed call @ref pm_notifier_unregister or
```
```suggestion
 * @note It is not allowed to call @ref pm_notifier_unregister or
```

erwango commented on include/power/power.h:
```diff
@@ -48,6 +48,10 @@ extern unsigned char pm_idle_exit_notify;
+ *
+ * @note It is not allowed call @ref pm_notifier_unregister or
+ *       @ref pm_notifier_register from these callbacks because they are called
+ *       with the spin locked in those functions locked.
```
```suggestion
 *       with the spin locked in those functions.
```

erwango commented on subsys/power/power.c:
```diff
@@ -72,7 +72,11 @@ __weak void pm_power_state_exit_post_ops(struct pm_state_info info)
+	 * The kernel expects that irqs are unlocked after this.
 	 */
+
+	irq_unlock(0);
```
Since it has to be done anyway, can it be moved to the functions were this function is called, just after calling it?
Otherwise, I'd suggest to modify the function description to add that `irq_unlock` should be done in this function.

ceolin replied:
This does not necessarily has to be done in this function, I can imagine a SoC that set `PRIMASK`, call `irq_unlock()` in `pm_power_state_set`  and unset `PRIMASK` before leave it and don't implement `pm_power_state_exit_post_ops`  or implement but doesn't need to restore interruption ... The fact is that currently, if one these functions are not implemented we have a problem because IRQs are not restored. I think these weak functions should be removed though.

erwango commented on doc/reference/power_management/index.rst:
```diff
@@ -124,6 +124,27 @@ have higher wake latencies. Following is a thorough list of available states:
+Power States Constraint
+=======================
+
+The power management subsystem allows different Zephyr components and
+applications to set constraints on various power states preventing the
+system to go these states. This can be used by devices when executing
+tasks in background to avoid the system to go to state where it would
+lose context. Constraints can be set, released and checked using the
+follow APIs:
```
Is this also valid in case of `pm_force_suspend_devices` is used ?
If not, maybe it should be also described here.

ceolin replied:
These are system constraint and not device constraints, right now they are not explicitly connected, I mean, devices don't look system constraints and the system don't look devices constraints. `pm_force_suspend_devices` suspend devices independently of the system constraint

erwango replied:
From my point of view (and from what I recall from our discussions) devices should also be able to put constraints.
As for instance a some device may accept to work in sleep or standby modes and we need to be able to assert that.
I'm about to push a new version of the PM changes for stm32 drivers, this will be the opportunity to discuss that.

ceolin replied:
yep, you're right, device can set constraints. Is this not clear in that text ? If it is not I can change it to be more clear.

erwango replied:
To me, what would deserve to be clarified is the impact vs `pm_force_suspend_devices`.

ceolin replied:
Sorry, but I don't see the point here. These are system constraints and `pm_force_suspend_devices` is orthogonal to this. I've cited there that devices can put constraints. From this API and system perspective, does not matter if it was a device or the application that set the constraint, the behavior is the same.

erwango replied:
Ok

erwango commented on doc/reference/power_management/index.rst:
```diff
@@ -124,6 +124,27 @@ have higher wake latencies. Following is a thorough list of available states:
+
+.. doxygenfunction:: pm_constraint_get
+   :project: Zephyr
+
```
Additionally, I think it could be useful to add a mention about `device_busy_foo` apis here, to explain the difference from  power manager point of view.

ceolin replied:
We need to update device power management documentation indeed, but right now I'm just looking system power management, is that ok ?

erwango replied:
Sure

Commit:
power: doc: Create a subgroup to constraint API

Create a separate sub group for constraint APIs. This way it is
possible to reference this api directly in the power management
reference without mess with the rest of the documentation.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: docs: sys_suspend -> pm_system_suspend

Change references to an old function name.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>
